cache/lib/cache_web/controllers/cas_controller.ex

```diff
@@ -3,14 +3,32 @@ defmodule CacheWeb.CASController do
 
   alias Cache.BodyReader
   alias Cache.Disk
+  alias Cache.S3
   alias Cache.S3UploadWorker
 
+  require Logger
+
   def download(conn, %{"id" => id, "account_handle" => account_handle, "project_handle" => project_handle}) do
-    internal_path = Disk.internal_accel_path(account_handle, project_handle, id)
+    key = Disk.cas_key(account_handle, project_handle, id)
 
-    conn
-    |> put_resp_header("x-accel-redirect", internal_path)
-    |> send_resp(:ok, "")
+    if Disk.exists?(account_handle, project_handle, id) do
+      local_path = Disk.local_accel_path(account_handle, project_handle, id)
+
+      conn
+      |> put_resp_header("x-accel-redirect", local_path)
+      |> send_resp(:ok, "")
+    else
+      case S3.presign_download_url(key) do
+        {:ok, url} ->
+          conn
+          |> put_resp_header("x-accel-redirect", S3.remote_accel_path(url))
+          |> send_resp(:ok, "")
+
+        {:error, _reason} ->
+          conn
+          |> send_resp(:not_found, "")
```

Comment:
this can either be a 404 or a valid error – do we at least want to report non-404 error if we don't pass it to the client?